registry/remote/repository.go

```diff
@@ -193,28 +193,40 @@ func (r *Repository) FetchReference(ctx context.Context, reference string) (ocis
 // and returns the parsed reference. If the parsed reference does not share
 // the same base reference with the Repository r, ParseReference returns a
 // wrapped error ErrInvalidReference.
-func (r *Repository) ParseReference(reference string) (registry.Reference, error) {
-	ref, err := registry.ParseReference(reference)
-	if err != nil {
-		// reference is not a FQDN
-		if index := strings.IndexByte(reference, '@'); index != -1 {
-			// drop tag since the digest is present
-			reference = reference[index+1:]
-		}
+func (r *Repository) ParseReference(reference string) (ref registry.Reference, err error) {
+	if ref, err = registry.ParseReference(reference); err != nil {
 		ref = registry.Reference{
 			Registry:   r.Reference.Registry,
 			Repository: r.Reference.Repository,
 			Reference:  reference,
 		}
-		if err = ref.ValidateReference(); err != nil {
+
+		// reference is not a FQDN
+		if index := strings.IndexByte(reference, '@'); index != -1 {
+			// drop tag since the digest is present
+			ref.Reference = reference[index+1:]
+			err = ref.ValidateReferenceAsDigest()
+		} else if index = strings.IndexByte(reference, ':'); index != -1 {
+			// no @ but :, so it can only be a digest
+			err = ref.ValidateReferenceAsDigest()
+		} else {
+			err = ref.ValidateReferenceAsTag()
+		}
+
+		if err != nil {
 			return registry.Reference{}, err
 		}
 	} else if ref.Registry != r.Reference.Registry || ref.Repository != r.Reference.Repository {
-		return registry.Reference{}, fmt.Errorf("%w %q: expect %q", errdef.ErrInvalidReference, ref, r.Reference)
+		return registry.Reference{}, fmt.Errorf(
+			"%w: mismatch between received `%q` and expected `%q`",
```

Comment:
`%q` is already double quoted.
```suggestion
			"%w: mismatch between received %q and expected %q",
```

Reply:
Fixed.